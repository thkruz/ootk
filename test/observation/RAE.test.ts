import {
  AngularDistanceMethod,
  DEG2RAD,
  Degrees,
  EpochUTC,
  J2000,
  Kilometers,
  KilometersPerSecond,
  RAE,
  Radians,
  Vector3D,
} from '../../src/main';

describe('RAE', () => {
  let exampleDate: Date;
  let rae: RAE;

  beforeEach(() => {
    exampleDate = new Date(1705109326817);
    rae = new RAE(
      EpochUTC.fromDateTime(exampleDate),
      1000 as Kilometers,
      10 * DEG2RAD as Radians,
      20 * DEG2RAD as Radians,
    );
  });

  it('should be constructable', () => {
    expect(rae).toMatchSnapshot();
  });

  it('should be constructable from degrees', () => {
    const raeDeg = RAE.fromDegrees(
      EpochUTC.fromDateTime(exampleDate),
      1000 as Kilometers,
      10 as Degrees,
      20 as Degrees,
    );

    expect(raeDeg).toMatchSnapshot();
    expect(raeDeg).toMatchObject(rae);
  });

  // azimuthDegrees
  it('should return the azimuth in degrees', () => {
    expect(rae.az).toMatchSnapshot();
  });

  // elevationDegrees
  it('should return the elevation in degrees', () => {
    expect(rae.el).toMatchSnapshot();
  });

  // azimuthRateDegrees
  it('should return the azimuth rate in degrees', () => {
    expect(rae.azRate).toMatchSnapshot();
  });

  // elevationRateDegrees
  it('should return the elevation rate in degrees', () => {
    expect(rae.elRate).toMatchSnapshot();
  });

  // toString
  it('should return a string representation', () => {
    expect(rae.toString()).toMatchSnapshot();
  });

  // fromStateVector
  it('should return a RAE object from a state vector', () => {
    const state = new J2000(
      EpochUTC.fromDateTime(exampleDate),
      new Vector3D<Kilometers>(7000 as Kilometers, 7000 as Kilometers, 8000 as Kilometers),
      new Vector3D<KilometersPerSecond>(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );
    const site = new J2000(
      EpochUTC.fromDateTime(exampleDate),
      new Vector3D<Kilometers>(6000 as Kilometers, 7000 as Kilometers, 8000 as Kilometers),
      new Vector3D<KilometersPerSecond>(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    const rae = RAE.fromStateVector(state, site);

    expect(rae).toMatchSnapshot();
  });

  // toStateVector
  it('should return a state vector', () => {
    const site = new J2000(
      EpochUTC.fromDateTime(exampleDate),
      new Vector3D<Kilometers>(6000 as Kilometers, 7000 as Kilometers, 8000 as Kilometers),
      new Vector3D<KilometersPerSecond>(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    expect(rae.toStateVector(site)).toMatchSnapshot();
  });

  // position
  it('should return the position vector', () => {
    const site = new J2000(
      EpochUTC.fromDateTime(exampleDate),
      new Vector3D<Kilometers>(6000 as Kilometers, 7000 as Kilometers, 8000 as Kilometers),
      new Vector3D<KilometersPerSecond>(0 as KilometersPerSecond, 0 as KilometersPerSecond, 0 as KilometersPerSecond),
    );

    expect(rae.position(site)).toMatchSnapshot();
  });

  // angle
  it('should return the angle', () => {
    const rae2 = new RAE(
      EpochUTC.fromDateTime(exampleDate),
      1000 as Kilometers,
      15 * DEG2RAD as Radians,
      25 * DEG2RAD as Radians,
    );

    expect(rae.angle(rae2)).toMatchSnapshot();
    expect(rae.angle(rae2, AngularDistanceMethod.Haversine)).toMatchSnapshot();
    expect(rae.angle(rae2, AngularDistanceMethod.Haversine))
      .not.toEqual(rae.angle(rae2, AngularDistanceMethod.Cosine));
  });

  // angleDegrees
  it('should return the angle in degrees', () => {
    const rae2 = new RAE(
      EpochUTC.fromDateTime(exampleDate),
      1000 as Kilometers,
      15 * DEG2RAD as Radians,
      25 * DEG2RAD as Radians,
    );

    expect(rae.angleDegrees(rae2)).toMatchSnapshot();
  });
});
