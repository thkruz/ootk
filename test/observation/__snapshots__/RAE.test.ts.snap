// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RAE should be constructable 1`] = `
RAE {
  "azRad": 0.17453292519943295,
  "azRateRad": undefined,
  "elRad": 0.3490658503988659,
  "elRateRad": undefined,
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "rng": 1000,
  "rngRate": undefined,
}
`;

exports[`RAE should be constructable from degrees 1`] = `
RAE {
  "azRad": 0.17453292519943295,
  "azRateRad": undefined,
  "elRad": 0.3490658503988659,
  "elRateRad": undefined,
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "rng": 1000,
  "rngRate": undefined,
}
`;

exports[`RAE should return a RAE object from a state vector 1`] = `
RAE {
  "azRad": 4.202021284209623,
  "azRateRad": 0.00043952100964508644,
  "elRad": 0.5130090849662382,
  "elRateRad": -0.0003354532635408476,
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "rng": 999.9999999999991,
  "rngRate": 0.5104272486229879,
}
`;

exports[`RAE should return a state vector 1`] = `
J2000 {
  "epoch": EpochUTC {
    "posix": 1705109326.817,
  },
  "position": Vector3D {
    "x": 5649.940123598527,
    "y": 6839.039773929244,
    "z": 8922.794608001712,
  },
  "velocity": Vector3D {
    "x": -0.4986876341007287,
    "y": 0.41048541483137624,
    "z": 0.0011464115053515227,
  },
}
`;

exports[`RAE should return a string representation 1`] = `
"[RazEl]
  Epoch:     2024-01-13T01:28:46.817Z
  Azimuth:   10.0000°
  Elevation: 20.0000°
  Range:     1000.000 km"
`;

exports[`RAE should return the angle 1`] = `0.11878030341845575`;

exports[`RAE should return the angle 2`] = `0.11878030341845577`;

exports[`RAE should return the angle in degrees 1`] = `6.805610075160859`;

exports[`RAE should return the azimuth in degrees 1`] = `10`;

exports[`RAE should return the azimuth rate in degrees 1`] = `undefined`;

exports[`RAE should return the elevation in degrees 1`] = `20`;

exports[`RAE should return the elevation rate in degrees 1`] = `undefined`;

exports[`RAE should return the position vector 1`] = `
Vector3D {
  "x": 5649.940123598527,
  "y": 6839.039773929244,
  "z": 8922.794608001712,
}
`;
